### DISCUSSION TITLE
Go to definition/reference missing model methods

### DISCUSSION THREAD

**Discussion by karambaq** [microsoft/pyright#5175] [Q&A] [answered]
I'm using neovim with pyright and working on Django project and it's not showing references/definition of the class methods sometimes, like this:

```python
    queryset = (
        models.Model.objects
        .actual() # <- Here i'm trying to see references or definition with `gr` or `gd` but nothing happens
        .visible()
```

If I'm trying to search by word under cursor it's ok, even LSP dynamic workspace symbols finding this method fine, but I'm wondering if pyright is capable of doing it by himself?

**Comment by erictraut** [comment]
I'm guessing that `models` in your code sample is a symbol imported from the `django.db` module? And I assume that you have the latest django installed without any django type stubs? I admit that I know nothing about django, but I can tell you what I see in the library from a static type perspective.

Django isn't a "py.typed" library, so it doesn't claim to supply any type information. That means pyright will assume all symbols imported from it are of type `Unknown` if you've set `useLibraryCodeForTypes` to `false`. If you've set `useLibraryCodeForTypes` to `true` (which is the default), pyright will attempt to infer types from the django source code. I'll assume that you have `useLibraryCodeForTypes` set to true.

The type of the expression `models.Model` is a class `Model` defined in the `django/db/models.py` file. I don't see any static definition for an attribute called `objects` in the `Model` class. Perhaps this attribute is dynamically injected into the object or class at runtime? If so, pyright won't have any visibility into it.

The `Model` class has a metaclass called `ModelBase`. If I install the `django-stubs` stub library, the the `ModelBase` metaclass includes a property called `objects`, but the property declaration indicates that it returns a value of type `BaseManager[Model]`. The `BaseManager` class doesn't have any method called `actual`.

```python
class ModelBase(type):
    ...
    @property
    def objects(cls: Type[_M]) -> BaseManager[_M]: ...
```

Based on the static type information available in the `django` library (and in the `django-stubs` stub library), I think pyright is doing the right thing here. I'm guessing that django's behavior is simply too dynamic for a static type analyzer to handle. There may be other tools that perform runtime analysis or introspection that can help you out.

**Reply by karambaq** [reply]
I apologize for confusing you and making you look for something that was not meant and spend your time. 

`Model` - is my model from models.py, I just changed it's name for security reasons, better would be if I called it `MyModel`

**Reply by erictraut** [reply]
Ah, then I don't think I have enough information to answer your question. Does your `MyModel` class define a class variable named `objects`? If so, have you supplied a type annotation for this variable?

**Reply by karambaq** [reply]
`objects` is the field that every django model have by default:
https://github.com/django/django/blob/main/django/db/models/base.py#L419

But sometimes It can be customised in model like this:
https://docs.djangoproject.com/en/3.2/topics/db/managers/#custom-managers

And I think I found corresponding topic on django-stubs repo:
https://github.com/typeddjango/django-stubs/pull/1025

**Reply by karambaq** [reply]
If you'll need more info let me know

**Reply by erictraut** [reply]
Have you provided a type annotation for `objects` in your class?

To provide additional assistance, I'd need a self-contained, minimal code sample that demonstrates the problem you're seeing.

**Reply by karambaq** [reply]
I think I got the case, It doesn't work when method under cursor is in the parent class of `objects` field class. And I found out that it doesn't work with simple classes, here is the minimal example:

```python
class SomeMixinClass:
    @staticmethod
    def method():
        ...


class SomeObjectsClass(SomeMixinClass):
    @staticmethod
    def another_method():
        ...


class SomeModel:
    objects = SomeObjectsClass


SomeModel.objects.another_method().method()
```

If I'm on `another_method` and pressing `gd` - all good, but if on just `method` - nothing happens

UPD: It seems strange but in Django each of "Manager" class methods (objects) returns Queryset so you can chain them
I can provide example with Django tomorrow.

**Reply by karambaq** [reply]
This is minimal example with Django:

```python3
from typing import Optional, List

from django.db import models


class ActualQuerySetMixin(models.QuerySet):
    def actual(self) -> models.QuerySet:
        return self


class MyModelQuerySet(ActualQuerySetMixin):
    ...


class MyModelManager(models.Manager):
    ...


class MyModel(models.Model):

    objects = MyModelManager.from_queryset(MyModelQuerySet)()


MyModel.objects.actual()

```

If you will stay on `actual` and will try to go to definition - nothing happens

**Reply by erictraut** [reply]
I'm once again assuming that you do not have the `django-stubs` stub library installed. (Can you please confirm?)

And once again, you have not provided a type annotation for `objects` as I suggested above. The type of `objects` cannot be properly inferred from the django source code in this case, so its type is `Unknown`. That means there's no way for the type checker to determine what symbol the `.actual` is referencing. If you provide a proper type annotation for `objects`, it should work fine.

**Reply by karambaq** [reply]
It's installed:

<img width="1378" alt="image" src="https://github.com/microsoft/pyright/assets/14233969/0979cbf2-3d45-4766-afa3-b2047ba54922">

I provided type annotation for objects, however it still doesn't find it definition and gives an error:

<img width="756" alt="image" src="https://github.com/microsoft/pyright/assets/14233969/bd7a1b32-7b58-4a21-baa2-3a1209a3bc98">

**Reply by erictraut** [reply]
Are you sure that's the right annotation? You've indicated that `objects` is of type `MyModelManager`, but this class doesn't have an `actual` method. Did you mean instead to indicate that its type is `MyModelQuerySet`, which has an `actual` method?

**Reply by karambaq** [reply]
Yes, you're right! That's looks strange and pyright don't like it

<img width="829" alt="image" src="https://github.com/microsoft/pyright/assets/14233969/c9d58dfd-3153-41ad-b87c-08d31eab171d">


But it works with `actual` method now, but it doesn't work for anything that "chains" after `actual()` method, for example `other` method is not found if it is called after `actual()`, I understand why, but is is possible to fix somehow?

```python
from django.db import models


class ActualQuerySetMixin(models.QuerySet):
    def actual(self) -> models.QuerySet:
        return self


class MyModelQuerySet(ActualQuerySetMixin):
    def other(self) -> models.QuerySet:
        return self


class MyModelManager(models.Manager):
    ...


class MyModel(models.Model):

    objects: MyModelQuerySet = MyModelManager.from_queryset(MyModelQuerySet)()


MyModel.objects.actual().other() # <- only `actual` is found
MyModel.objects.other().actual() # <- only `other` is found
MyModel.objects.other() # <- fine
MyModel.objects.actual() # <- fine

```

**Reply by karambaq** [reply]
Unfortunately didn't find a solution yet

**Reply by erictraut** [reply]
Do you have type checking enabled? If I paste the above code into my editor with type checking enabled, pyright emits errors that indicate that `other` and `actual` are not known members of the `QuerySet` class. Because of these errors, you cannot "go to the definition" of these symbols because pyright has no knowledge of them.

**Reply by karambaq** [reply]
Yes, I see this errors too, my only question is how to fix them and is it possible

**Reply by erictraut** [reply, marked as answer]
You need to use the correct type annotations in your code. In your code above, `ActualQuerySetMixin.actual` is annotated to return an object of type `models.QuerySet`. I think that what you want here is `typing.Self`. For details about the `Self` type, refer to [PEP 673](https://peps.python.org/pep-0673/).

**Reply by karambaq** [reply]
Thank you!
